maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,253 @@
+                buffer_time=buffer_time
+            )
+            pending_decision_evt = self._event_buffer.gen_decision_event(evt.tick, payload=decision_payload)
+            evt.add_immediate_event(pending_decision_evt)
+            self._pending_action_vm_id = vm_req.vm_id
+        else:
+            # Postpone the buffer duration ticks by config setting.
+            if buffer_time > 0:
+                postpone_payload = payload
+                postpone_payload.buffer_time -= self._delay_duration
+                self._total_latency.latency_due_to_resource += self._delay_duration
+                postpone_evt = self._event_buffer.gen_cascade_event(
+                    evt.tick + self._delay_duration, Events.REQUIREMENTS, payload=postpone_payload)
+                self._event_buffer.insert_event(postpone_evt)
+            else:
+                # Fail
+                # TODO Implement failure logic.
+                self._failed_requirements += 1
+
+    def _on_vm_finished(self, evt: AtomEvent):
+        """Callback when there is a VM in the end cycle."""
+        # Get the end-cycle VM info.
+        payload: VmFinishedPayload = evt.payload
+        vm_id = payload.vm_id
+        virtual_machine: VirtualMachine = self._vm[vm_id]
+        util_vm_cores = virtual_machine.util_cpu * virtual_machine.req_cpu
+
+        # Release PM resources.
+        physical_machine: PhysicalMachine = self._machines[virtual_machine.pm_id]
+        physical_machine.req_cpu -= virtual_machine.req_cpu
+        physical_machine.req_mem -= virtual_machine.req_mem
+        physical_machine.util_cpu = (
+            (physical_machine.cap_cpu * physical_machine.util_cpu / 100 - util_vm_cores)
+            * 100 / physical_machine.cap_cpu
+        )
+        physical_machine.remove_vm(vm_id)
+
+        # Remove dead VM.
+        self._vm.pop(vm_id)
+
+        # VM allocation succeed.
+        self._success_requirements += 1
+
+    def _on_action_received(self, evt: CascadeEvent):
+        """Callback wen we get an action from agent."""
+        cur_tick: int = evt.tick
+        action: Action = evt.payload
+        assign: bool = action.assign
+        virtual_machine: VirtualMachine = action.vm_req
+
+        if virtual_machine.id != self._pending_action_vm_id:
+            print("The VM id sent by agent is invalid.")
+
+        if assign:
+            pm_id = action.pm_id
+            lifetime = virtual_machine.lifetime
+            # Update VM information.
+            virtual_machine.pm_id = pm_id
+            virtual_machine.start_tick = cur_tick
+            virtual_machine.end_tick = cur_tick + lifetime
+            # The index of the VM's utilization in util_series.
+            vm_util_index = cur_tick - virtual_machine.start_tick
+            cur_util_vm_cores = virtual_machine.util_series[vm_util_index] * virtual_machine.req_cpu / 100
+            virtual_machine.util_cpu = cur_util_vm_cores / virtual_machine.req_cpu * 100
```

Comment:
It seems that line 224 and line 225 can be combined to a single line:
``vm.util_cpu = vm.get_util(cur_tick=cur_tick)``